dlp/snippets/inspect: TestInspectGCS failed

This test failed!

To configure my behavior, see [the Build Cop Bot documentation](https://github.com/googleapis/repo-automation-bots/tree/master/packages/buildcop).

If I'm commenting on this issue too often, add the `buildcop: quiet` label and
I will stop commenting.

---

commit: cb5809cc565ef7a694d55d158d7f37bb60aa22f3
buildURL: [Build Status](https://source.cloud.google.com/results/invocations/ac04e57d-4891-4388-b64a-ffca164d7720), [Sponge](http://sponge2/ac04e57d-4891-4388-b64a-ffca164d7720)
status: failed
<details><summary>Test output</summary><br><pre>inspect_test.go:106: inspectGCSFile(fake_ssn.txt) got err: Receive: draining</pre></details>

The logs aren't very verbose, unless I'm missing something. Here's all I could find:

```
--- FAIL: TestInspectGCS (0.30s)
    --- FAIL: TestInspectGCS/fake_ssn.txt (62.09s)
        inspect_test.go:106: inspectGCSFile(fake_ssn.txt) got err: Receive: draining
```

Just looking at the code, though, I see a few potential issues that are likely to cause flakes.

1. The timeout on the pubsub operation is only 1 minute. It's common for the API call to take about that long, but that leaves no margin for error. For comparison, the equivalent Java sample for Inspect GCS uses a 15 minute timeout waiting for the pubsub message. 
2. The tests all run using exactly the same bucket, topic, and subscription. If more than one test is running concurrently, they are likely to interfere with each other. Most of the other language tests use some sort of guid suffix to make sure that anything written to is unique to the test run.

--- a/dlp/snippets/inspect/inspect_bigquery.go
+++ b/dlp/snippets/inspect/inspect_bigquery.go
@@ -161,9 +161,9 @@ func inspectBigquery(w io.Writer, projectID string, infoTypeNames []string, cust
 	fmt.Fprintf(w, "Created job: %v\n", j.GetName())
 
 	// Wait for the inspect job to finish by waiting for a PubSub message.
-	// This only waits for 1 minute. For long jobs, consider using a truly
+	// This only waits for 10 minutes. For long jobs, consider using a truly
 	// asynchronous execution model such as Cloud Functions.
-	ctx, cancel := context.WithTimeout(ctx, 1*time.Minute)
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Minute)
 	defer cancel()
 	err = s.Receive(ctx, func(ctx context.Context, msg *pubsub.Message) {
 		// If this is the wrong job, do not process the result.

--- a/dlp/snippets/inspect/inspect_datastore.go
+++ b/dlp/snippets/inspect/inspect_datastore.go
@@ -161,9 +161,9 @@ func inspectDatastore(w io.Writer, projectID string, infoTypeNames []string, cus
 	fmt.Fprintf(w, "Created job: %v\n", j.GetName())
 
 	// Wait for the inspect job to finish by waiting for a PubSub message.
-	// This only waits for 1 minute. For long jobs, consider using a truly
+	// This only waits for 10 minutes. For long jobs, consider using a truly
 	// asynchronous execution model such as Cloud Functions.
-	ctx, cancel := context.WithTimeout(ctx, 1*time.Minute)
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Minute)
 	defer cancel()
 	err = s.Receive(ctx, func(ctx context.Context, msg *pubsub.Message) {
 		// If this is the wrong job, do not process the result.

--- a/dlp/snippets/inspect/inspect_gcs.go
+++ b/dlp/snippets/inspect/inspect_gcs.go
@@ -157,9 +157,9 @@ func inspectGCSFile(w io.Writer, projectID string, infoTypeNames []string, custo
 	fmt.Fprintf(w, "Created job: %v\n", j.GetName())
 
 	// Wait for the inspect job to finish by waiting for a PubSub message.
-	// This only waits for 1 minute. For long jobs, consider using a truly
+	// This only waits for 10 minutes. For long jobs, consider using a truly
 	// asynchronous execution model such as Cloud Functions.
-	ctx, cancel := context.WithTimeout(ctx, 1*time.Minute)
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Minute)
 	defer cancel()
 	err = s.Receive(ctx, func(ctx context.Context, msg *pubsub.Message) {
 		// If this is the wrong job, do not process the result.

--- a/dlp/snippets/inspect/inspect_test.go
+++ b/dlp/snippets/inspect/inspect_test.go
@@ -24,11 +24,12 @@ import (
 	"cloud.google.com/go/datastore"
 	"cloud.google.com/go/storage"
 	"github.com/GoogleCloudPlatform/golang-samples/internal/testutil"
+	"github.com/gofrs/uuid"
 )
 
 const (
-	topicName        = "dlp-inspect-test-topic"
-	subscriptionName = "dlp-inspect-test-sub"
+	topicName        = "dlp-inspect-test-topic-"
+	subscriptionName = "dlp-inspect-test-sub-"
 
 	ssnFileName = "fake_ssn.txt"
 	bucketName  = "golang-samples-dlp-test"
@@ -49,8 +50,9 @@ func TestInspectDatastore(t *testing.T) {
 	for _, test := range tests {
 		t.Run(test.kind, func(t *testing.T) {
 			t.Parallel()
+			u := uuid.Must(uuid.NewV4()).String()[:8]
 			buf := new(bytes.Buffer)
-			if err := inspectDatastore(buf, tc.ProjectID, []string{"US_SOCIAL_SECURITY_NUMBER"}, []string{}, []string{}, topicName, subscriptionName, tc.ProjectID, "", test.kind); err != nil {
+			if err := inspectDatastore(buf, tc.ProjectID, []string{"US_SOCIAL_SECURITY_NUMBER"}, []string{}, []string{}, topicName+u, subscriptionName+u, tc.ProjectID, "", test.kind); err != nil {
 				t.Errorf("inspectDatastore(%s) got err: %v", test.kind, err)
 			}
 			if got := buf.String(); !strings.Contains(got, test.want) {
@@ -101,8 +103,9 @@ func TestInspectGCS(t *testing.T) {
 	for _, test := range tests {
 		t.Run(test.fileName, func(t *testing.T) {
 			t.Parallel()
+			u := uuid.Must(uuid.NewV4()).String()[:8]
 			buf := new(bytes.Buffer)
-			if err := inspectGCSFile(buf, tc.ProjectID, []string{"US_SOCIAL_SECURITY_NUMBER"}, []string{}, []string{}, topicName, subscriptionName, bucketName, test.fileName); err != nil {
+			if err := inspectGCSFile(buf, tc.ProjectID, []string{"US_SOCIAL_SECURITY_NUMBER"}, []string{}, []string{}, topicName+u, subscriptionName+u, bucketName, test.fileName); err != nil {
 				t.Errorf("inspectGCSFile(%s) got err: %v", test.fileName, err)
 			}
 			if got := buf.String(); !strings.Contains(got, test.want) {
@@ -256,8 +259,9 @@ func TestInspectBigquery(t *testing.T) {
 	for _, test := range tests {
 		t.Run(test.table, func(t *testing.T) {
 			t.Parallel()
+			u := uuid.Must(uuid.NewV4()).String()[:8]
 			buf := new(bytes.Buffer)
-			if err := inspectBigquery(buf, tc.ProjectID, []string{"US_SOCIAL_SECURITY_NUMBER"}, []string{}, []string{}, topicName, subscriptionName, tc.ProjectID, bqDatasetID, test.table); err != nil {
+			if err := inspectBigquery(buf, tc.ProjectID, []string{"US_SOCIAL_SECURITY_NUMBER"}, []string{}, []string{}, topicName+u, subscriptionName+u, tc.ProjectID, bqDatasetID, test.table); err != nil {
 				t.Errorf("inspectBigquery(%s) got err: %v", test.table, err)
 			}
 			if got := buf.String(); !strings.Contains(got, test.want) {

--- a/dlp/snippets/risk/categorical.go
+++ b/dlp/snippets/risk/categorical.go
@@ -98,9 +98,9 @@ func riskCategorical(w io.Writer, projectID, dataProject, pubSubTopic, pubSubSub
 	fmt.Fprintf(w, "Created job: %v\n", j.GetName())
 
 	// Wait for the risk job to finish by waiting for a PubSub message.
-	// This only waits for 1 minute. For long jobs, consider using a truly
+	// This only waits for 10 minutes. For long jobs, consider using a truly
 	// asynchronous execution model such as Cloud Functions.
-	ctx, cancel := context.WithTimeout(ctx, 1*time.Minute)
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Minute)
 	defer cancel()
 	err = s.Receive(ctx, func(ctx context.Context, msg *pubsub.Message) {
 		// If this is the wrong job, do not process the result.

--- a/dlp/snippets/risk/k_anonymity.go
+++ b/dlp/snippets/risk/k_anonymity.go
@@ -104,9 +104,9 @@ func riskKAnonymity(w io.Writer, projectID, dataProject, pubSubTopic, pubSubSub,
 	fmt.Fprintf(w, "Created job: %v\n", j.GetName())
 
 	// Wait for the risk job to finish by waiting for a PubSub message.
-	// This only waits for 1 minute. For long jobs, consider using a truly
+	// This only waits for 10 minutes. For long jobs, consider using a truly
 	// asynchronous execution model such as Cloud Functions.
-	ctx, cancel := context.WithTimeout(ctx, 1*time.Minute)
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Minute)
 	defer cancel()
 	err = s.Receive(ctx, func(ctx context.Context, msg *pubsub.Message) {
 		// If this is the wrong job, do not process the result.

--- a/dlp/snippets/risk/k_map.go
+++ b/dlp/snippets/risk/k_map.go
@@ -114,9 +114,9 @@ func riskKMap(w io.Writer, projectID, dataProject, pubSubTopic, pubSubSub, datas
 	fmt.Fprintf(w, "Created job: %v\n", j.GetName())
 
 	// Wait for the risk job to finish by waiting for a PubSub message.
-	// This only waits for 1 minute. For long jobs, consider using a truly
+	// This only waits for 10 minutes. For long jobs, consider using a truly
 	// asynchronous execution model such as Cloud Functions.
-	ctx, cancel := context.WithTimeout(ctx, 1*time.Minute)
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Minute)
 	defer cancel()
 	err = s.Receive(ctx, func(ctx context.Context, msg *pubsub.Message) {
 		// If this is the wrong job, do not process the result.

--- a/dlp/snippets/risk/l_diversity.go
+++ b/dlp/snippets/risk/l_diversity.go
@@ -108,9 +108,9 @@ func riskLDiversity(w io.Writer, projectID, dataProject, pubSubTopic, pubSubSub,
 	fmt.Fprintf(w, "Created job: %v\n", j.GetName())
 
 	// Wait for the risk job to finish by waiting for a PubSub message.
-	// This only waits for 1 minute. For long jobs, consider using a truly
+	// This only waits for 10 minutes. For long jobs, consider using a truly
 	// asynchronous execution model such as Cloud Functions.
-	ctx, cancel := context.WithTimeout(ctx, 1*time.Minute)
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Minute)
 	defer cancel()
 	err = s.Receive(ctx, func(ctx context.Context, msg *pubsub.Message) {
 		// If this is the wrong job, do not process the result.

--- a/dlp/snippets/risk/numerical.go
+++ b/dlp/snippets/risk/numerical.go
@@ -98,9 +98,9 @@ func riskNumerical(w io.Writer, projectID, dataProject, pubSubTopic, pubSubSub, 
 	fmt.Fprintf(w, "Created job: %v\n", j.GetName())
 
 	// Wait for the risk job to finish by waiting for a PubSub message.
-	// This only waits for 1 minute. For long jobs, consider using a truly
+	// This only waits for 10 minutes. For long jobs, consider using a truly
 	// asynchronous execution model such as Cloud Functions.
-	ctx, cancel := context.WithTimeout(ctx, 1*time.Minute)
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Minute)
 	defer cancel()
 	err = s.Receive(ctx, func(ctx context.Context, msg *pubsub.Message) {
 		// If this is the wrong job, do not process the result.

--- a/dlp/snippets/risk/risk_test.go
+++ b/dlp/snippets/risk/risk_test.go
@@ -17,20 +17,27 @@ package risk
 
 import (
 	"bytes"
+	"context"
 	"strings"
 	"testing"
 	"time"
 
+	"cloud.google.com/go/pubsub"
 	"github.com/GoogleCloudPlatform/golang-samples/internal/testutil"
+	"github.com/gofrs/uuid"
 )
 
 const (
-	riskTopicName        = "dlp-risk-test-topic"
-	riskSubscriptionName = "dlp-risk-test-sub"
+	riskTopicName        = "dlp-risk-test-topic-"
+	riskSubscriptionName = "dlp-risk-test-sub-"
 )
 
 func TestRisk(t *testing.T) {
 	tc := testutil.SystemTest(t)
+	client, err := pubsub.NewClient(context.Background(), tc.ProjectID)
+	if err != nil {
+		t.Fatalf("pubsub.NewClient: %v", err)
+	}
 	tests := []struct {
 		name string
 		fn   func(r *testutil.R)
@@ -39,7 +46,9 @@ func TestRisk(t *testing.T) {
 			name: "Numerical",
 			fn: func(r *testutil.R) {
 				buf := new(bytes.Buffer)
-				err := riskNumerical(buf, tc.ProjectID, "bigquery-public-data", riskTopicName, riskSubscriptionName, "nhtsa_traffic_fatalities", "accident_2015", "state_number")
+				u := uuid.Must(uuid.NewV4()).String()[:8]
+				err := riskNumerical(buf, tc.ProjectID, "bigquery-public-data", riskTopicName+u, riskSubscriptionName+u, "nhtsa_traffic_fatalities", "accident_2015", "state_number")
+				defer cleanupPubsub(t, client, riskTopicName+u, riskSubscriptionName+u)
 				if err != nil {
 					r.Errorf("riskNumerical got err: %v", err)
 					return
@@ -53,7 +62,9 @@ func TestRisk(t *testing.T) {
 			name: "Categorical",
 			fn: func(r *testutil.R) {
 				buf := new(bytes.Buffer)
-				err := riskCategorical(buf, tc.ProjectID, "bigquery-public-data", riskTopicName, riskSubscriptionName, "nhtsa_traffic_fatalities", "accident_2015", "state_number")
+				u := uuid.Must(uuid.NewV4()).String()[:8]
+				err := riskCategorical(buf, tc.ProjectID, "bigquery-public-data", riskTopicName+u, riskSubscriptionName+u, "nhtsa_traffic_fatalities", "accident_2015", "state_number")
+				defer cleanupPubsub(t, client, riskTopicName+u, riskSubscriptionName+u)
 				if err != nil {
 					r.Errorf("riskCategorical got err: %v", err)
 					return
@@ -67,7 +78,9 @@ func TestRisk(t *testing.T) {
 			name: "K Anonymity",
 			fn: func(r *testutil.R) {
 				buf := new(bytes.Buffer)
-				err := riskKAnonymity(buf, tc.ProjectID, "bigquery-public-data", riskTopicName, riskSubscriptionName, "nhtsa_traffic_fatalities", "accident_2015", "state_number", "county")
+				u := uuid.Must(uuid.NewV4()).String()[:8]
+				err := riskKAnonymity(buf, tc.ProjectID, "bigquery-public-data", riskTopicName+u, riskSubscriptionName+u, "nhtsa_traffic_fatalities", "accident_2015", "state_number", "county")
+				defer cleanupPubsub(t, client, riskTopicName+u, riskSubscriptionName+u)
 				if err != nil {
 					r.Errorf("riskKAnonymity got err: %v", err)
 					return
@@ -81,7 +94,9 @@ func TestRisk(t *testing.T) {
 			name: "L Diversity",
 			fn: func(r *testutil.R) {
 				buf := new(bytes.Buffer)
-				err := riskLDiversity(buf, tc.ProjectID, "bigquery-public-data", riskTopicName, riskSubscriptionName, "nhtsa_traffic_fatalities", "accident_2015", "city", "state_number", "county")
+				u := uuid.Must(uuid.NewV4()).String()[:8]
+				err := riskLDiversity(buf, tc.ProjectID, "bigquery-public-data", riskTopicName+u, riskSubscriptionName+u, "nhtsa_traffic_fatalities", "accident_2015", "city", "state_number", "county")
+				defer cleanupPubsub(t, client, riskTopicName+u, riskSubscriptionName+u)
 				if err != nil {
 					r.Errorf("riskLDiversity got err: %v", err)
 					return
@@ -95,7 +110,9 @@ func TestRisk(t *testing.T) {
 			name: "K Map",
 			fn: func(r *testutil.R) {
 				buf := new(bytes.Buffer)
-				riskKMap(buf, tc.ProjectID, "bigquery-public-data", riskTopicName, riskSubscriptionName, "san_francisco", "bikeshare_trips", "US", "zip_code")
+				u := uuid.Must(uuid.NewV4()).String()[:8]
+				riskKMap(buf, tc.ProjectID, "bigquery-public-data", riskTopicName+u, riskSubscriptionName+u, "san_francisco", "bikeshare_trips", "US", "zip_code")
+				defer cleanupPubsub(t, client, riskTopicName+u, riskSubscriptionName+u)
 				if got, want := buf.String(), "Created job"; !strings.Contains(got, want) {
 					r.Errorf("riskKMap got %s, want substring %q", got, want)
 				}
@@ -108,5 +125,28 @@ func TestRisk(t *testing.T) {
 			t.Parallel()
 			testutil.Retry(t, 20, 2*time.Second, test.fn)
 		})
+	}
+}
+
+func cleanupPubsub(t *testing.T, client *pubsub.Client, topicName, subName string) {
+	ctx := context.Background()
+	topic := client.Topic(topicName)
+	if exists, err := topic.Exists(ctx); err != nil {
+		t.Logf("Exists: %v", err)
+		return
+	} else if exists {
+		if err := topic.Delete(ctx); err != nil {
+			t.Logf("Delete: %v", err)
+		}
+	}
+
+	s := client.Subscription(subName)
+	if exists, err := s.Exists(ctx); err != nil {
+		t.Logf("Exists: %v", err)
+		return
+	} else if exists {
+		if err := s.Delete(ctx); err != nil {
+			t.Logf("Delete: %v", err)
+		}
 	}
 }

